### PR TITLE
channel_filter: return GraphErrors instead of FailedPluginExecution

### PR DESCRIPTION
**Pull request by vrutkovs**
MissingParams error is returned if channel is not set, InvalidParams error - if channel does not match the regexp.

Ref: https://jira.coreos.com/browse/CIN-69

```
$ curl -s --header 'Accept:application/json' http://localhost:8081/v1/graph
{"kind":"missing_params","value":"mandatory client parameters missing: could not find key 'channel'"}
$ curl -s --header 'Accept:application/json' http://localhost:8081/v1/graph\?channel\=invalid:channel
{"kind":"invalid_params","value":"invalid client parameters: channel 'invalid:channel' does not match regex '^[0-9a-z\\-\\.]+$'"}
```

### CODE REVIEW

**Comment by lucab** [comment]
ok to test

File: cincinnati/src/plugins/internal/channel_filter.rs [hunk 1]
```diff
@@ -56,14 +57,17 @@ impl InternalPlugin for ChannelFilterPlugin {
             self.key_suffix.to_owned(),
         ))
         .and_then(|(internal_io, key_prefix, key_suffix)| {
-            let channel = get_multiple_values!(internal_io.parameters, "channel")?.clone();
+            let channel = match get_multiple_values!(internal_io.parameters, "channel"){
+                Ok(c) => c.clone(),
+                Err(e) => return Err(GraphError::MissingParams(vec!(e.to_string()))),
```
Comment (lucab) [inline cincinnati/src/plugins/internal/channel_filter.rs]: This should be `vec!["channel".to_string()]`.

File: cincinnati/src/plugins/internal/channel_filter.rs [hunk 2]
```diff
@@ -56,14 +57,17 @@ impl InternalPlugin for ChannelFilterPlugin {
             self.key_suffix.to_owned(),
         ))
         .and_then(|(internal_io, key_prefix, key_suffix)| {
-            let channel = get_multiple_values!(internal_io.parameters, "channel")?.clone();
+            let channel = match get_multiple_values!(internal_io.parameters, "channel"){
```
Comment (lucab) [inline cincinnati/src/plugins/internal/channel_filter.rs]: I suspect we could directly fix the macro to return the proper error instead of `bail!()`, and adjust its name to make it explicitly scoped to client parameters. @steveej?
Sidenote: for the single value case here, this should be effectively equal to a `Option::ok_or()`.

File: cincinnati/src/plugins/internal/channel_filter.rs [hunk 3]
```diff
@@ -104,6 +108,10 @@ impl InternalPlugin for ChannelFilterPlugin {
                 graph,
                 parameters: internal_io.parameters,
             })
+        })
+        .map_err(move |e| {
+            error!("error filtering graph: {}", e);
```
Comment (lucab) [inline cincinnati/src/plugins/internal/channel_filter.rs]: Policy-engine should be already logging this error, can you confirm?

**Review by lucab — COMMENTED** [review]
Good patch! I have just a few minor comments.

File: cincinnati/src/plugins/internal/channel_filter.rs (hunk 1 above)
Comment (vrutkovs) [reply cincinnati/src/plugins/internal/channel_filter.rs]: I was trying to avoid hardcoding "channel" in yet another place, but sure, I'll change that

File: cincinnati/src/plugins/internal/channel_filter.rs (hunk 2 above)
Comment (vrutkovs) [reply cincinnati/src/plugins/internal/channel_filter.rs]: I'd rather wrap the error in the plugin so that this macros could be reused elsewhere

File: cincinnati/src/plugins/internal/channel_filter.rs (hunk 1 above)
Comment (lucab) [reply cincinnati/src/plugins/internal/channel_filter.rs]: Sorry, let me rephrase for clarity: the vector should contain only the keynames of missing parameters, not full human-friendly error strings.

File: cincinnati/src/plugins/internal/channel_filter.rs (hunk 3 above)
Comment (vrutkovs) [reply cincinnati/src/plugins/internal/channel_filter.rs]: Right, this is being ignored - removed

File: cincinnati/src/plugins/internal/channel_filter.rs (hunk 2 above)
Comment (steveeJ) [reply cincinnati/src/plugins/internal/channel_filter.rs]: How about this:


```diff
diff --git a/cincinnati/src/plugins/internal/channel_filter.rs b/cincinnati/src/plugins/internal/channel_filter.rs
index 87231eb..c26dac5 100644
--- a/cincinnati/src/plugins/internal/channel_filter.rs
+++ b/cincinnati/src/plugins/internal/channel_filter.rs
@@ -57,10 +57,9 @@ impl InternalPlugin for ChannelFilterPlugin {
             self.key_suffix.to_owned(),
         ))
         .and_then(|(internal_io, key_prefix, key_suffix)| {
-            let channel = match get_multiple_values!(internal_io.parameters, "channel"){
-                Ok(c) => c.clone(),
-                Err(_) => return Err(GraphError::MissingParams(vec!["channel".to_string()])),
-            };
+            let channel = get_multiple_values!(internal_io.parameters, "channel")
+                .map_err(|e| GraphError::MissingParams(vec![e.to_string()]))?
+                .clone();

             if !CHANNEL_VALIDATION_REGEX_RE.is_match(&channel) {
                 return Err(GraphError::InvalidParams(format!(
diff --git a/cincinnati/src/plugins/macros.rs b/cincinnati/src/plugins/macros.rs
index 2cc216c..1911a64 100644
--- a/cincinnati/src/plugins/macros.rs
+++ b/cincinnati/src/plugins/macros.rs
@@ -9,7 +9,7 @@ macro_rules! get_multiple_values {
                     if let Some(value) = $map.get($key) {
                         value
                     } else {
-                        bail!("could not find key '{}'", $key)
+                        bail!("{}", $key)
                     },
                 )
             };
@@ -25,7 +25,7 @@ macro_rules! get_multiple_values {
                             if let Some(value) = $map.get($key) {
                                 value
                             } else {
-                                bail!("could not find key '{}'", $key)
+                                bail!("{}", $key)
                             },
                         )*
                     )
```

Comment (steveej) [reply cincinnati/src/plugins/internal/channel_filter.rs]: How about this:


```diff
diff --git a/cincinnati/src/plugins/internal/channel_filter.rs b/cincinnati/src/plugins/internal/channel_filter.rs
index 87231eb..c26dac5 100644
--- a/cincinnati/src/plugins/internal/channel_filter.rs
+++ b/cincinnati/src/plugins/internal/channel_filter.rs
@@ -57,10 +57,9 @@ impl InternalPlugin for ChannelFilterPlugin {
             self.key_suffix.to_owned(),
         ))
         .and_then(|(internal_io, key_prefix, key_suffix)| {
-            let channel = match get_multiple_values!(internal_io.parameters, "channel"){
-                Ok(c) => c.clone(),
-                Err(_) => return Err(GraphError::MissingParams(vec!["channel".to_string()])),
-            };
+            let channel = get_multiple_values!(internal_io.parameters, "channel")
+                .map_err(|e| GraphError::MissingParams(vec![e.to_string()]))?
+                .clone();

             if !CHANNEL_VALIDATION_REGEX_RE.is_match(&channel) {
                 return Err(GraphError::InvalidParams(format!(
diff --git a/cincinnati/src/plugins/macros.rs b/cincinnati/src/plugins/macros.rs
index 2cc216c..1911a64 100644
--- a/cincinnati/src/plugins/macros.rs
+++ b/cincinnati/src/plugins/macros.rs
@@ -9,7 +9,7 @@ macro_rules! get_multiple_values {
                     if let Some(value) = $map.get($key) {
                         value
                     } else {
-                        bail!("could not find key '{}'", $key)
+                        bail!("{}", $key)
                     },
                 )
             };
@@ -25,7 +25,7 @@ macro_rules! get_multiple_values {
                             if let Some(value) = $map.get($key) {
                                 value
                             } else {
-                                bail!("could not find key '{}'", $key)
+                                bail!("{}", $key)
                             },
                         )*
                     )
```

**Review by steveej — CHANGES_REQUESTED** [review]
I added one comment as requested by @lucab. While trying it out locally I also noted that the code hasn't been run through rustfmt which we usually do.

**Comment by steveej** [comment]
Thanks!

/lgtm

/hold
In case @lucab has any final remarks.

**Review by steveej — APPROVED** [review]
